guppylang/diagnostic.py

```diff
@@ -109,3 +151,222 @@ class Error(Diagnostic, Protocol):
+
+        ```
+           |
+        42 | def foo(x: blah) -> None:
+           |            ---- Non-primary span label
+        ```
+
+        Optionally includes up to `prefix_lines` preceding source lines to give
+        additional context.
+        """
+        # Check how much space we need to reserve for the leading line numbers
+        ll_length = len(str(span.end.line))
+        highlight_char = "^" if is_primary else "-"
+
+        def render_line(line: str, line_number: int | None = None) -> None:
+            """Helper method to render a line with the line number bar on the left."""
+            if line_number is not None:
+                ll = str(line_number)
+                self.buffer.append(ll + " " * (ll_length - len(ll)) + " | " + line)
+            else:
+                self.buffer.append(" " * ll_length + " | " + line)
+
+        # One line of padding
+        render_line("")
+
+        # Grab all lines we want to display and remove excessive leading whitespace
+        prefix_lines = min(prefix_lines, span.start.line - 1)
+        all_lines = self.source.span_lines(span, prefix_lines)
+        leading_whitespace = min(len(line) - len(line.lstrip()) for line in all_lines)
+        if leading_whitespace > self.MAX_LEADING_WHITESPACE:
+            remove = leading_whitespace - self.OPTIMAL_LEADING_WHITESPACE
+            all_lines = [line[remove:] for line in all_lines]
+            span = span.shift_left(remove)
+
+        # Render prefix lines
+        for i, line in enumerate(all_lines[:prefix_lines]):
+            render_line(line, span.start.line - prefix_lines + i)
+        span_lines = all_lines[prefix_lines:]
+
+        if span.is_multiline:
+            [first, *middle, last] = span_lines
+            render_line(first, span.start.line)
+            # Compute the subspan that only covers the first line and render it's
+            # highlight banner
+            first_span = Span(span.start, Loc(span.file, span.start.line, len(first)))
+            first_highlight = " " * first_span.start.column + highlight_char * len(
+                first_span
+            )
+            render_line(first_highlight)
+            # Omit everything in the middle
+            if middle:
+                render_line("...")
+            # The last line is handled uniformly with the single-line case below.
+            # Therefore, create a subspan that only covers the last line.
+            last_span = Span(Loc(span.file, span.end.line, 0), span.end)
+        else:
+            [last] = span_lines
+            last_span = span
+
+        # Render the last span line and add highlights
+        render_line(last, span.end.line)
+        last_highlight = " " * last_span.start.column + highlight_char * len(last_span)
+
+        # Render the label next to the highlight
+        if label:
+            [label_first, *label_rest] = textwrap.wrap(
+                label,
+                self.MAX_LABEL_LINE_LEN,
```

Comment:
So this is the length of the label AFTER the highlight bar. Irrespective of how much space there is left on the line after said highlight bar. How about instead you ask textwap to layout `highlight_char * len(last_span) + " " + label` with initial indent `last_span.start.column` and subsequent indents, hmmm, you might want to go *backwards* ?
```
           ^^^^ there is |
    some problem here    |
    that won't fit on the|
    rest of the line
```
(`|` to indicate line end boundary, not printed)

Think if your `test_long_label` has very little space after the `==`...

Reply:
I think the current alignment looks nicer, but I agree that the spacing might become problematic. 

In a previous version I implemented a layout algorithm that abbreviates the start/end of the line and/or the middle of the span with `...` ro ensure that the `^^^` doesn't go too far to the right and everything fits into a specified width

```
   |
42 | ...else foo(looo...ng_id).more_stuff(...
   |             ^^^^^^^^^^^^ Span label with
   |                          line break
```

I think that is the best solution but would have made this PR 150-200 lines longer 😅

I created #556 as a followup